cpp/src/arrow/util/bit_block_counter.h

```diff
@@ -244,6 +244,30 @@ class ARROW_EXPORT OptionalBinaryBitBlockCounter {
     }
   }
 
+  BitBlockCount NextOrNotBlock() {
```

Comment:
I'm duplicating code here. `NextOrBlock` is identical to 
https://github.com/apache/arrow/blob/bd107272c076687212720a10514e7698eff25ffe/cpp/src/arrow/util/bit_block_counter.h#L223-L231
except for `case HasBitmap::BOTH:` where `NextOrNotWord` is called. Is there a good way to avoid the repetition?

Reply:
I think it's ok for now, we can revisit later if desired.